src/main/java/org/truffleruby/core/proc/ProcNodes.java

```diff
@@ -274,6 +276,37 @@ protected RubyProc createSameArityProc(RubyProc userProc, RubyProc block) {
         }
     }
 
+    @Primitive(name = "proc_specify_arity", lowerFixnum = 1)
+    public abstract static class ProcSpecifyArityNode extends PrimitiveArrayArgumentsNode {
+        @Specialization
+        protected RubyProc specifyArityProc(RubyProc block, int argc) {
+            Arity oldArity = block.sharedMethodInfo.getArity();
+            final Arity newArity;
+            if (argc <= -1) {
+                newArity = new Arity(0, oldArity.getOptional(), true);
+            } else {
+                newArity = new Arity(argc, oldArity.getOptional(), false);
+            }
+
+            SharedMethodInfo newSharedMethodInfo = block.sharedMethodInfo.createWithNewArity(newArity);
+            final RubyProc composedProc = new RubyProc(
+                    coreLibrary().procClass,
+                    getLanguage().procShape,
+                    block.type,
+                    newSharedMethodInfo,
+                    block.callTargets,
+                    block.callTarget,
+                    block.declarationFrame,
+                    block.declarationVariables,
+                    block.method,
+                    block.block,
+                    block.frameOnStackMarker,
+                    block.declarationContext);
```

Comment:
Sorry I know this constructor is a bit of a joke now it's so long!

Reply:
It's fine, `proc_create_same_arity` does the same.
It seems a good idea to refactor to have a `withArity(newArity)` method on RubyProc too to make this nicer and use both in `proc_create_same_arity` and here.